Refactor runtime composition context and type matching logic

sinkingsugar opened the pull request:
- Replace `inherited` map in `InternalCompositionContext` with a `LayeredMap` for better layer management.
- Adjust `validateConnection` and `internalComposeWire` to use `LayeredMap`.
- Add nested iterator to `LayeredMap` for traversing layers efficiently.
- Ensure proper handling of key insertions and deletions across layers.
- Refactor `matchTypes` call for improved readability.
- Remove unused constructor in `CompositionContext`.


<!--
Before submitting your PR, please review the following checklist:

- [ ] **DO NOT** submit a PR without having discussed the change first in a related issue. If none exist, create one first.
- [ ] **CONSIDER** adding a unit test if your PR resolves an issue.
- [ ] **DO** keep pull requests small so they can be easily reviewed.
- [ ] **DO** make sure all unit tests pass.
- [ ] **DO** make sure not to introduce any new compiler warnings.
- [ ] **AVOID** breaking the continuous integration build.
- [ ] **AVOID** making significant changes to the overall architecture.
-->

sinkingsugar committed:
Refactor runtime composition context and type matching logic

- Replace `inherited` map in `InternalCompositionContext` with a `LayeredMap` for better layer management.
- Adjust `validateConnection` and `internalComposeWire` to use `LayeredMap`.
- Add nested iterator to `LayeredMap` for traversing layers efficiently.
- Ensure proper handling of key insertions and deletions across layers.
- Refactor `matchTypes` call for improved readability.
- Remove unused constructor in `CompositionContext`.

sinkingsugar committed:
Refactor LayeredMap to support blocking and improve iteration

Refactored LayeredMap to introduce blocking layers and enhanced iterator behavior. Updated internal iteration logic and size calculation to account for blocker tags. Improved begin and end iterator logic to handle blocking layers. Added globals map to CompositionContext for broader scope handling.

sinkingsugar committed:
Enhance variable lookup and validation in runtime

- Adjusted `findVariable` to exit early on pure wires with a `break` instead of `return nullptr`.
- Included `globals` in `sharedStorage` during validation to prioritize exposed variables.
- Added global variable lookup fallback in `validateConnection`.
- Improved error messages for clarity, including exposed variable types and contexts (direct, inherited, globals).
- Refactored `internalComposeWire` to handle `globals` distinctly from `inherited` variables.
- Improved `Push` error handling with more descriptive messages for mutable, protected, and tracked variables.

sinkingsugar committed:
Merge commit '434fcc4587f8bc0f72cd8083204da41d4774e4e5' into gio/compose-refactor

sinkingsugar committed:
Add isolating flag to SHInstanceData for scoped composition

Introduce an `isolating` flag in `SHInstanceData` to enable scoped composition handling. Adjust inheritance logic in `internalComposeWire` to respect this flag, ensuring shared context variables are isolated when required. Modify `insert` methods to support forced insertion into the last layer when isolating. Update `Isolate` shard to set the `isolating` flag during composition.

sinkingsugar committed:
Refactor(core): remove globals and improve context isolation

Refactor runtime composition by removing the `globals` map and enhancing the usage of `inherited` for shared context. Eliminate the `isolating` flag and ensure `privateContext` is nullified to create new contexts as needed. Simplify shared storage handling and streamline variable resolution logic.

guusw commented on shards/core/runtime.hpp:
```diff
@@ -246,6 +246,222 @@ struct DefaultHelpText {
+template <typename K, typename V> class LayeredMap {
+  std::vector<std::unordered_map<K, V>> layers;
+  std::vector<bool> blocker_tags;
+  size_t first_blocker_index = std::numeric_limits<size_t>::max(); // Index of the first blocker
```
What does this do? shouldn't searching from back to front and stop on blocker tag work for all cases?

sinkingsugar replied:
Ah eh, didn't check, blame AI 🤖🤣

sinkingsugar replied:
I think its supposed to cache the index of the first layer but probably redundant, I'll double check

sinkingsugar committed:
Refactor runtime and LayeredMap logic for clarity and efficiency

- Replaced `USE_TSAN` with `USE_RUST_TSAN` in Rust sanitizer checks.
- Updated `pushLayer` and `popLayer` logic to remove unnecessary blocker index tracking.
- Simplified `insert`, `erase`, `find`, `size`, and `begin` methods in `LayeredMap`.
- Improved handling of blocked layers by iterating over layers directly.
- Removed redundant code and optimized method implementations for better readability and maintainability.

sinkingsugar committed:
Add composeV2 support and improve LayeredMap functionality

Extend shard composition with a new optional `composeV2` function, enabling enhanced flexibility during validation and composition phases. Introduce a `LayeredMap` implementation for managing variable scope with blocker tags and nested layers, improving composition context inheritance. Update affected shards to use `composeV2` where appropriate and streamline type checking, error handling, and inheritance logic. Include related updates to runtime and shard macros for full integration.

sinkingsugar committed:
Improve variable handling, formatting, and gfx shard logic

- Fix layer pop logic to properly update blocker tags.
- Adjust total size calculation in blocker-tag-aware loops.
- Avoid unnecessary variable capture for pure wires in BaseRunner.
- Refactor formatting for gfx shard declarations and parameter descriptions.
- Ensure private context reset in generator wire composition.
- Cleanup shared data in FeatureShard properly and ensure readability.

sinkingsugar committed:
Refactor(runtime): expose findVariable function in runtime API

Expose the findVariable function in the runtime API by removing its inline specifier and declaring it in the runtime.hpp header. This allows external access to the function for locating variables in the context.

sinkingsugar committed:
Update imaging.cpp to increment output version correctly

Ensure the output version is incremented relative to the input version in Convolve, PremultiplyAlpha, and LoadImage structs to maintain consistency and track changes accurately.

sinkingsugar committed:
Refactor: migrate shards to composeV2 for consistency

Update shards to use the new `composeV2` method for improved type safety and shared variable access. Replace legacy `compose` calls with `composeV2` and ensure all relevant shards handle private context validation. Simplify type resolution logic and remove redundant loops for cleaner implementation.

sinkingsugar committed:
Refactor compose logic and improve type derivation

Refactor compose logic by introducing `composeVersion` to distinguish behaviors and enhance `deriveTypeInfo` with better error handling and profiling optimizations. Update relevant functions to use layered maps for variable lookups and ensure compatibility with new composition logic. Cleanup profiling-related code and ensure code consistency across modules.

sinkingsugar committed:
Merge commit 'f49b4dacb3649a0f7426183139074ee8c6a719b6' into gio/compose-refactor

sinkingsugar committed:
Refactor: improve error message formatting in validateConnection

Replace manual stringstream-based error message construction with a concise `fmt::format` implementation for better readability and maintainability.

guusw commented on shards/core/runtime.cpp:
```diff
@@ -1122,19 +1132,16 @@ SHComposeResult internalComposeWire(const std::vector<Shard *> &wire, SHInstance
   }
 
   if (data.shared.elements) {
-#if SH_EXTENDED_COMPOSE_PROFILING
```
Don't want to keep these for profiling?

guusw commented on shards/core/runtime.cpp:
```diff
@@ -1025,17 +1034,18 @@ void validateConnection(InternalCompositionContext &ctx) {
     }
 
     if (!matching) {
-      std::string error;
```
Can you change back to the shorter format (the fmt one)

guusw replied:
It basically just doesn't log all the exposed variables and makes the error message a bit more concise

guusw commented on shards/core/type_info.cpp:
```diff
@@ -131,20 +133,38 @@ SHTypeInfo deriveTypeInfo(const SHVar &value, const SHInstanceData &data, std::v
-        std::string_view name(info.name);
-        if (name == sv) {
-          expInfo->push_back(SHExposedTypeInfo{.name = info.name, .exposedType = info.exposedType});
+      if (data.composeVersion == 2) {
```
Shouldn't this also be allowed on compose v1, since it should still have the privateContext, just filled shared, etc.

sinkingsugar replied:
true

guusw commented on shards/modules/imaging/imaging.cpp:
```diff
@@ -745,7 +747,7 @@ struct LoadImage {
     }
 
-    _output.version = 0;
+    _output.version = input.version + 1;
```
Input is bytes here though

guusw replied:
Should be 0 I think

sinkingsugar committed:
Refactor: remove `composeVersion` and improve `LayeredMap`

Remove unused `composeVersion` field from `SHInstanceData` and related references. Update `LayeredMap` to allow modification of existing entries and refine iterator behavior. Simplify `deriveTypeInfo` logic by removing outdated `composeVersion` handling. Fix `LoadImage` output version reset to zero. Clean up unnecessary whitespace and formatting inconsistencies.

sinkingsugar committed:
Refactor(core): streamline variable collection APIs

Replace exposed types with SHInstanceData in collectRequiredVariables and related functions. Update function signatures and calls across modules to improve clarity and consistency in handling exposed context variables. Move the definition of collectRequiredVariables to runtime.cpp for better modularization.

guusw committed:
Fix kdtree test
